src/parallel_parameter_search/walk_engine_optimization.py

```diff
@@ -0,0 +1,230 @@
+        engine_param_dict = {}
+
+        def add(name, min_value, max_value):
+            engine_param_dict[name] = trial.suggest_uniform(name, min_value, max_value)
+
+        def fix(name, value):
+            engine_param_dict[name] = value
+            trial.set_user_attr(name, value)
+
+        add('double_support_ratio', 0.0, 0.5)
+        add('freq', 0.5, 5)
+
+        add('foot_distance', foot_distance[0], foot_distance[1])
+        add('trunk_height', trunk_height[0], trunk_height[1])
+
+        add('trunk_phase', -0.5, 0.5)
+        add('trunk_swing', 0.0, 1.0)
+        add('trunk_z_movement', 0, z_movement)
+
+        add('trunk_x_offset', -trunk_x, trunk_x)
+        add('trunk_y_offset', -trunk_x, trunk_x)
+
+        add('trunk_pitch_p_coef_forward', -5, 5)
+        add('trunk_pitch_p_coef_turn', -5, 5)
+
+        add('trunk_pitch', -0.5, 0.5)
+        # fix('trunk_pitch', 0.0)
+        add('foot_rise', 0.05, 0.15)
+        # fix('foot_rise', foot_rise)
+
+        add('first_step_swing_factor', 0.0, 2)
+        # fix('first_step_swing_factor', 1)
+        fix('first_step_trunk_phase', -0.5)
+
+        # add('foot_overshoot_phase', 0.0, 1.0)
+        # add('foot_overshoot_ratio', 0.0, 1.0)
+        fix('foot_overshoot_phase', 1)
+        fix('foot_overshoot_ratio', 0.0)
+
+        fix('foot_apex_phase', 0.5)
+
+        # add('foot_z_pause', 0, 1)
+        # add('foot_put_down_phase', 0, 1)
+        # add('trunk_pause', 0, 1)
+        fix('foot_z_pause', 0)
+        fix('foot_put_down_phase', 1)
+        fix('trunk_pause', 0)
+
+        node_param_dict = {}
+        # walk engine should update at same speed as simulation
+        node_param_dict["engine_freq"] = 1 / self.sim.get_timestep()
+        # don't use loop closure when optimizing parameter
+        node_param_dict["pressure_phase_reset_active"] = False
+        node_param_dict["effort_phase_reset_active"] = False
+        node_param_dict["phase_rest_active"] = False
+        # make sure that steps are not limited
+        node_param_dict["imu_active"] = False
+        node_param_dict["max_step_x"] = 100.0
+        node_param_dict["max_step_y"] = 100.0
+        node_param_dict["max_step_xy"] = 100.0
+        node_param_dict["max_step_z"] = 100.0
+        node_param_dict["max_step_angular"] = 100.0
+
+        if self.walk_as_node:
+            self.set_params(engine_param_dict)
+            self.set_params(node_param_dict)
+        else:
+            self.current_params = engine_param_dict
+            self.walk.set_engine_dyn_reconf(engine_param_dict)
+            self.walk.set_node_dyn_reconf(node_param_dict)
+
+        # necessary for correct reset
+        self.trunk_height = self.current_params["trunk_height"]
+        self.trunk_pitch = self.current_params["trunk_pitch"]
+        try:
+            self.trunk_pitch_p_coef_forward = self.current_params["trunk_pitch_p_coef_forward"]
```

Comment:
```suggestion
            self.trunk_pitch_p_coef_forward = self.current_params.get("trunk_pitch_p_coef_forward", 0)
```
would be better instead of try/except